[inductor][scheduler] reorder scheduler nodes after fusion to reduce peak memory

xuanzhang816 opened the pull request:
**Motivations**: 
A topological order of the scheduler nodes that optimize the liveness of buffers can reduce the peak memory utilization. This has been observed and studied e.g., [here](https://arxiv.org/pdf/1910.02653) and [here](https://proceedings.mlr.press/v202/steiner23a/steiner23a.pdf).

**Solutions**:
1. implement a peak memory estimator via liveness analysis
2. implement a few memory aware topological sorting algorithms and pick the one with the lowest peak memory

**Results**: 
On some models we can reduce the peak memory significantly:
|             model             | batch size | peak_memory baseline | peak_memory new | ratio |
|:-----------------------------:|:----------:|:--------------------:|:---------------:|:-----:|
| alexnet                       | 128        |         1.17         |       0.99      | 1.19  |
| vgg16                         | 64         |         4.10         |       3.57      | 1.15  |
| DebertaV2ForQuestionAnswering | 1          |         11.60        |      10.56      | 1.10  |

In the presence of compiler based AC, peak memory can be further reduced:
|              model             | batch size | peak_memory baseline | peak_memory new | ratio |
|:------------------------------:|:----------:|:--------------------:|:---------------:|:-----:|
| AlbertForMaskedLM              | 4          |         6.87         |       6.43      | 1.07  |
| AlbertForQuestionAnswering     | 4          |         8.69         |       7.76      | 1.12  |
| MobileBertForQuestionAnswering | 128        |         4.67         |       3.90      | 1.20  |

[Here](https://fb.workplace.com/groups/1075192433118967/posts/1499920537312819/?comment_id=1499938843977655&reply_comment_id=1499951630643043) is an internal use case.

**Other infos:**
* neutral model runtime, because the the reordering happens after fusion. So memory saving is _for free_.
* minimal compile time overhead as the algorithm is linear in the number of edges of the inductor graph. For all hugglingface benchmark models, the additional compile time is less than 1 second.
* no peak memory regression since we only adopt a new order if the peak memory is reduced based on the estimator. However, the model is unaware of operators' working memories, but for large models, the working memory should be negligible. We haven't observed any significant regressions on all of our tests.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

xuanzhang816 committed:
reorder for peak memory

yf225 commented on torch/_inductor/config.py:
```diff
@@ -266,6 +266,9 @@ def autotune_remote_cache_default() -> Optional[bool]:
 ]
 
+# enable operator reordering for peak memory optimization
+reorder_for_peak_memory = os.environ.get("TORCHINDUCTOR_REORDER_PEAK_MEMORY") == "1"
```
nit
```suggestion
reorder_for_peak_memory = os.environ.get("TORCHINDUCTOR_REORDER_FOR_PEAK_MEMORY") == "1"
```

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -2105,6 +2169,45 @@ def add_user(
+            for input_buf_name, input_buf in self.name_to_input_buf.items():
+                input_buf.set_users(name_to_users[input_buf_name].items)
+
+    def compute_size_for_scheduler_buffer(self) -> None:
```
I think it could be worth moving all the new functions to a dedicated file (e.g. `memory.py` in the same folder), to make the `scheduler.py` file shorter and also provide an easy place to find all the memory passes.

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -166,6 +200,14 @@ class BaseSchedulerNode:
+    index: int
+    memory_to_free: int
+    size: int
+    measure: int
```
Do these new new fields have to live in `BaseSchedulerNode` object (and in `SchedulerBuffer` object), or can we use dictionaries to represent them? Reason is that they are only used in the reorder for peak memory pass and would be great to make them self-contained within the pass.

xuanzhang816 replied:
Adding them here to deal with mypy linter complains. But it seems that once I moved the changes to a separate file memory.py, this complaint is gone. So will remove them from here.

xuanzhang816 replied:
nvm, the complaint is still there and it just took mypy a while to pick them up.

xuanzhang816 replied:
I set up two classes to record these information, one called `MemoryPlanningInfoForBuffer` and one called `MemoryPlanningInfoForNode`. Putting them in `scheduler.py` right now instead of `memory.py` to avoid circular imports.

yf225 reviewed (commented):
thanks! leaving some comments regarding organization

yf225 commented on torch/_inductor/config.py:
```diff
@@ -266,6 +266,9 @@ def autotune_remote_cache_default() -> Optional[bool]:
 ]
 
+# enable operator reordering for peak memory optimization
+reorder_for_peak_memory = os.environ.get("TORCHINDUCTOR_REORDER_PEAK_MEMORY") == "1"
```
we might want to add unit tests as well to prevent regression, e.g. example: https://github.com/pytorch/pytorch/blob/main/test/distributed/test_compute_comm_reordering.py

xuanzhang816 replied:
added a unittest. Let me know if more detailed ones are needed.

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -2142,8 +2245,323 @@ def can_eliminate_user(user: NodeUser) -> bool:
+            cur_memory += memory[t]
+            max_memory = max(max_memory, cur_memory)
+
+        return max_memory
```
curious: is there a way to also get a dictionary of "node_index -> cur_memory at that index" from this function above? (maybe we can return it as a dict?) I believe it will be useful for FSDP memory-based prefetching as well

xuanzhang816 replied:
I think the `cur_memory` should give this information, let me return both max_memory and `cur_memory` for this function.

xuanzhang816 replied:
The function now returns both peak memory and memories at each step.

xuanzhang816 committed:
reorganize code as suggested

xuanzhang816 committed:
remove debug prints

xuanzhang816 committed:
add logging

xuanzhang816 committed:
add unittest

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -1783,7 +1815,7 @@ def _init(self, nodes: List[ir.Operation]) -> None:
         self.mutation_renames: Dict[str, str] = {}
 
-        self.compute_dependencies()
+        name_to_users = self.compute_dependencies()
```
Curious: is this change because we need a snapshot of `name_to_users` that's unchanged by the immediately following steps?

xuanzhang816 replied:
Ooops, this is actually no longer needed after some changes. I forgot to remove this.

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -66,17 +67,47 @@
+    index: int = 0
+    memory_to_free: int = 0
+    size: int = 0
+    measure: int = 0
```
maybe we can import these two dataclasses from memory.py as well?

xuanzhang816 replied:
I can't believe how much time I spent on this, fixing linter and circular import issues 😅 but finally made it work. I moved these two classes to the `memory.py` file now.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,525 @@
+from .virtualized import V
+
+
+torch_log = logging.getLogger("torch")
```
`log = logging.getLogger(__name__)` might be better here and also more similar to how scheduler.py uses logging

xuanzhang816 replied:
will fix.

yf225 commented on test/inductor/test_memory.py:
```diff
@@ -0,0 +1,83 @@
+        )
+        # check for correctness
+        outp = compiled_model(inp)
+        self.assertTrue(same(outp, outp_corr))
```
do we need to add detailed unit tests to check that each memory optimization algorithm is working correctly?

xuanzhang816 replied:
Yes!!!! Would the best practice be having a setting in config such as `peak_memory_optimization_algo` and then patch them individually in the unittest?

yf225 replied:
I think we can have a list of algorithm method names (or function handles) as argument to the `reorder_for_peak_memory()` method (and have the default value be the current list of algorithms). And then in the unit test, we can monkey-patch the `reorder_for_peak_memory()` method to just take in the specific algorithm name (something similar to https://github.com/pytorch/pytorch/blob/386884e5534bc812e4f90dcc94d420e148f20f2b/test/distributed/_composable/fsdp/test_fully_shard_compile.py#L205-L212)

xuanzhang816 replied:
Added a few more unittests, one for each sorting methods.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,525 @@
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    while nodes_to_schedule:
```
I am slightly worried about the while loop here being an infinite loop if the graph is ill-constructed (it probably doesn't happen right now, but probably important to prevent it from happening since we might want to enable it by default). Wonder is there a way to add a constraint for "maximum number of iterations" (can be a function of the graph size), and then just exit from this algorithm if it's not solved within the maximum iterations?

xuanzhang816 replied:
Added a check for the number of iterations for the while loops. If it is longer than expected, will raise a runtime error. Also added a "try - except" clause for each run in the main algorithm.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,525 @@
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    while nodes_to_schedule:
```
same here :)

xuanzhang816 replied:
I think I will also add a `verify_order` function to make sure the returned order is correct.

xuanzhang816 committed:
reorg code and add more unittests

xuanzhang816 committed:
fix a bug in bfs

xuanzhang816 committed:
add guard for while loops

xuanzhang816 committed:
raise runtime error for long while loops

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,533 @@
+
+def topological_sort_bfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    A BFS topological sort that selects nodes whose depednecies are executed
```
nit:
```suggestion
    A BFS topological sort that selects nodes whose dependencies are executed
```

xuanzhang816 committed:
fixed typo

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,533 @@
+
+
+@dataclasses.dataclass
+class InputBuffer:
```
I believe there is another `InputBuffer` class in ir.py: https://github.com/pytorch/pytorch/blob/68a7246f13e8e2afd55bcf21bad6634de414b404/torch/_inductor/ir.py#L3518 and it might be slightly hard to differentiate between the two during code search. Wonder is there a way to rename this dataclass to avoid the name duplication?

xuanzhang816 replied:
I changed the name to `FreeableInputBuffer` now.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,533 @@
+    graph_outputs: Set[str],
+) -> List[BaseSchedulerNode]:
+    """
+    A bfs-based greedy topological order.
```
it might be nice to have explanation of "lpmf" which is "Lowest Peak Memory First" (correct me if I am wrong 🙂)

xuanzhang816 replied:
Good point. That is correct, I added an explanation in the comments here.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,533 @@
+def topological_sort_bfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    A BFS topological sort that selects nodes whose dependencies are executed
+    the earliest. This follows a FIFO idea.
```
would you like to add more details on how this algorithm works? (would be great if we can have similar level of detail as in `topological_sort_lpmf`)

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,533 @@
+
+def topological_sort_dfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    Ensure nodes is in topologically sorted order
```
would you like to add more details on how this algorithm works, and some intuitive understanding on why it can lower peak memory? (would be great if we can have similar level of detail as in `topological_sort_lpmf`)

xuanzhang816 committed:
change name of InputBuffer, add comments

yf225 commented on test/inductor/test_memory.py:
```diff
@@ -0,0 +1,205 @@
+# Owner(s): ["module: inductor"]
```
We might need to add this test file to this list as well:
https://github.com/pytorch/pytorch/blob/1dfa07e885dbd86f14030648d80517de384c6c6b/.ci/pytorch/test.sh#L1398-L1406

yf225 replied:
also I wasn't able to find the new tests in the logs of a CI job yet... I spot-checked some of the jobs but couldn't find it :( we probably need to confirm it shows up in any job log to prove that it's running in CI

yf225 replied:
the easiest way is probably to intentionally introduce an error in the unit test, and see if the CI fails as expected..

xuanzhang816 replied:
Here is a screenshot showing that the purposely introduced error is indeed picked up by CI 

![image](https://github.com/user-attachments/assets/c6e3b152-451b-47ea-9dc9-868640875692)

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,543 @@
+    """
+    A BFS topological sort that selects nodes whose dependencies are executed the
+    earliest. This follows a FIFO idea. Specifically, at every iteration, for each node
+    that is schedulable, we gather the time when its predecessors nodes are executed,
+    and this sorted list of execution times of predecessors nodes defines the priority.
```
we might want to emphasize it's execution order and not execution time (which could be confused with length of execution):
```suggestion
    that is schedulable, we gather the order in which its predecessor nodes are executed,
    and this sorted list of execution orders of predecessor nodes defines the priority.
```

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,543 @@
+    index: int = 0
+    memory_to_free: int = 0
+    size: int = 0
+    measure: int = 0
```
is `measure` only used in the DFS algorithm? wonder if we can make this field internal to that algorithm function.

also is there a way to add more details on the meaning of `measure`? (the name might be a bit vague I think)

xuanzhang816 replied:
I changed the name from `measure` to `size_with_reads` to make it more clear.

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,543 @@
+class MemoryPlanningInfoForBuffer:
+    size_alloc: int = 0
+    size_free: int = 0
+    outdegree: int = 0
```
is `outdegree` only used in LPMF? wonder if we can make it internal to that algorithm function

yf225 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,543 @@
+    succ_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    indegree: int = 0
+    index: int = 0
+    memory_to_free: int = 0
```
is `memory_to_free` only used in LPMF? wonder if we can make it internal to that algorithm function

xuanzhang816 replied:
I was doing so before. So instead of declaring `memory_to_free` as an attribute in this dataclass, I dynamically set it inside the `_lpmf()` function. However, mypy linter is throwing errors. I am not sure how to not pre-declare these and also make linter happy :( We can discuss more about this.

yf225 replied:
sounds great - maybe a comment to call out it's lpmf only is sufficient

yf225 commented on torch/_inductor/scheduler.py:
```diff
@@ -77,6 +78,9 @@ class SchedulerBuffer:
     node: ir.Buffer
     defining_op: BaseSchedulerNode
     users: List[NodeUser] = dataclasses.field(default_factory=list)
+    mpi: MemoryPlanningInfoForBuffer = dataclasses.field(
```
I believe `.mpi` can mean two things in this PR: `MemoryPlanningInfoForBuffer` or `MemoryPlanningInfoForNode`. I wonder is there a way to easily differentiate the two? (maybe we can call them `mpi_buffer` and `mpi_node`?)

xuanzhang816 replied:
changed to mpi_buffer and mpi_node depending on the type.

xuanzhang816 committed:
check unittest on ci

xuanzhang816 committed:
fix unittest, distinguish mpi

xuanzhang816 committed:
fix dfs

yf225 reviewed (approved):
Thanks a lot!

xuanzhang816 committed:
adding comments for mpi attributes

xuanzhang816 committed:
log.info to log.warning

xuanzhang816 commented:
@pytorchbot merge

xuanzhang816 commented:
@pytorchbot merge -f "unrelated failures"

robert-hardwick commented:
This PR has caused CI failures in linux-aarch64 workflow. I've raised an issue #136464

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+            name_to_buf_info[sched_buf.get_name()] = BufferInfo(
+                sched_buf,
+                sched_buf.mpi_buffer.size_alloc,
+                sched_buf.mpi_buffer.size_free,
+                t,
+                t,
+            )
```
IMO, It makes it a little more difficult to follow to have intermediary states which are not correct. Here, the end_time is not right for a period of time. I would lean more towards having intermediary objects as you need them instead of incrementally updating objects such that there is a period of time where they are not correct. Kind of similar also to the dataclasses comment.

xuanzhang816 replied:
I restructured the code here to get rid of these "intermediary states". (these are interesting comments, I need to think a bit about how to address similar comments for dataclasses as quite some restructuring is needed)

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+        succ_node_time = [
+            node_name_to_time[succ_node.get_name()]
+            for succ_node in buf_info.buffer.mpi_buffer.succ_nodes
+            if succ_node.get_name() in node_name_to_time
```
When do we expect `succ_node.get_name() in node_name_to_time` to not be in `node_name_to_time` ? If there is an expected invariant where it shouldn't be, maybe can we assert that instead or error, instead of potentially skipping

xuanzhang816 replied:
you are right, removing this. You are making me realize that I have the habit of checking for existence whenever I do a dictionary lookup. :/

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+            for succ_node in buf_info.buffer.mpi_buffer.succ_nodes
+            if succ_node.get_name() in node_name_to_time
+        ]
+        if succ_node_time:
+            buf_info.end_time = max(succ_node_time)
```
when would a `succ_node_time` not be set ? if its the graph output case, can we just move that logic into here so we always have an end time.

xuanzhang816 replied:
There are two cases and one is the graph output case. The other case is when there is a vertical fusion -- e.g., assume `node1` has `buf1` as an output, and `buf1` is only consumed by `node2`, and assume `node1` and `node2` are fused, then `buf1` has no successor node.

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+
+    for node in nodes:
+        node.mpi_node.pred_nodes = list(
+            {
```
nit: set, determinism

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+                for pred_buffer in node.mpi_node.pred_buffers
+                if (
+                    isinstance(pred_buffer, SchedulerBuffer)
+                    and pred_buffer.defining_op.get_name() in name_to_fused_node
```
same comment about when would this not be present

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    graph_outputs: Set[str],
+) -> List[BaseSchedulerNode]:
+    """
+    A bfs-based greedy topological order. LPMF stands for "Least Peak Memory First".
```
does this come from a paper ? if so, link ?

xuanzhang816 replied:
Added.

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: Set[BaseSchedulerNode] = set()
```
OrderedSet

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi_node.indegree = len(node.mpi_node.pred_nodes)
+        if node.mpi_node.indegree == 0:
```
Like other comments, if we are going to inplace update the class, it would have the not yet computed state be explicitly represented (None), instead of a value which is a valid state later. It makes it more clear as we are following the logic / debugging.

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    output_memory = sum(
+        name_to_buf[buf_name].mpi_buffer.size_free
+        for buf_name in graph_outputs
+        if buf_name in name_to_buf
```
when is this not set ?

xuanzhang816 replied:
Good question. For the forward pass, `graph_outputs` can have things like `primals_*` but won't be in `name_to_buf` because they are not schedulerNodes' output buffers.

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+        node.mpi_node.size = sum(
+            buffer.mpi_buffer.size_alloc for buffer in node.get_outputs()
+        )
+        node.mpi_node.memory_to_free = 0
```
why do we re-assign this ?

xuanzhang816 replied:
removed this.

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    num_iters: int = 0
+    while num_iters < len(nodes) and nodes_to_schedule:
+        # select a node to schedule:
+        selected_node = min(
+            nodes_to_schedule,
+            key=lambda node: (
+                max(live_memory + node.mpi_node.size, max_memory),
+                node.mpi_node.size - node.mpi_node.memory_to_free,
+                node.mpi_node.index,
+            ),
+        )
```
Could we use a heap here ? instead of sorting on each iteration. for elements which are changed you can remove and reinsert

xuanzhang816 replied:
I actually thought about using a heap here but `live_memory` is updating from iteration to iteration, `node.mpi_node.memory_to_free` could be changing from iteration to iteration since how much memory can be freed after scheduling a node depends on what has been sceduled fore.  

So if we were to use a heap all elements in the heap would need to be updated from iteration to iteration

eellison replied:
I guess the reason this isn't as bad is because it's O(n) * O(max_concurrently_schedulable_nodes).

But, just to discuss - the live memory is the same for every node. If `live_memory + node.mpi_node.size  < max_memory`, does it actually matter what we schedule ? I'm not sure we need the max here: `max(live_memory + node.mpi_node.size, max_memory)`. 

I think you could conceivably avoid updating the live memory and just update the memory_to_free ? anyways not needed but thinking aloud.

eellison reviewed (commented):
Very cool ! Awesome to memory improvements already internally. Would be great if you could do a benchmark run, let me know if you're not aware how to. There were no major issues, just smaller things. 

I would be curious if you have done any testing of the various methods. How they compare in terms of compilation time or memory and if any of the methods was strictly worse than the others. 

I didnt get to the last few methods yet but submitting this review anyway and will do others in follow up.

xuanzhang816 commented:
@eellison 

### benchmark run

[Here](https://lookaside.facebook.com/intern/diff/file/data/?number=1888625565&download=1) is the spreadsheet with all benchmark models from torchbench, timm, and huggingface. 
* I run each model twice or three times, and use the memory number from the last run to compare. 
* test two settings -- (1) default (2) compiler based auto activation checkpointing with memory budget set to 0.4.

### compilation time

Timed and logged the runtime for huggingface models and uploaded the results [here](https://www.internalfb.com/phabricator/paste/view/P1596116802). The additional compile time is less than 1 second for all models; and much less than 1 second for most models.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+
+@dataclasses.dataclass
+class MemoryPlanningInfoForBuffer:
+    size_alloc: int = 0
+    size_free: int = 0
```
Do you have code pointers for symints? Seems that all size related computation within shceduler are based on size hints.

eellison replied:
Yea, that's correct - we just take size hints...

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    # compute the size of SchedulerBuffer with MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
+        if isinstance(sched_buf.node.layout, MultiOutputLayout):
+            sched_buf.mpi_buffer = MemoryPlanningInfoForBuffer()
```
yep, removing the re-instantiation.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+) -> None:
+    """
+    For scheduling and memory estimation, for each scheduler node, we maintain
+    a list of its dependency buffers (SchedulerBuffer and FreeableInputBuffer).
```
Yes! Let me modify the comment to be more clear.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    pred_buffers: List[Union[SchedulerBuffer, FreeableInputBuffer]] = dataclasses.field(
+        default_factory=list
+    )
+    pred_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
+    succ_nodes: List[BaseSchedulerNode] = dataclasses.field(default_factory=list)
```
You are right. Changing these to OrderedSet.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+            sched_buf.mpi_buffer.size_free = sched_buf.mpi_buffer.size_alloc
+
+    # compute the size of SchedulerBuffer with MultiOutputLayout layout
+    for sched_buf in name_to_buf.values():
```
I need to process the size for SchedulerBuffer without MultiOutputLayout layout first before doing so for those with MultiOutputLayout layout. So that's why I kept two separate loops.

xuanzhang816 commented:
@eellison I am addressing the comments locally one by one and resolving them here to keep track of which one I have done. I will create a new PR after I go through everything.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    """
+    for node in nodes:
+        node.mpi_node = MemoryPlanningInfoForNode()
+        node.mpi_node.pred_buffers = []
```
making sure I understand the line you are referring to. I removed the line `node.mpi_node.pred_buffers = []`. The line `node.mpi_node = MemoryPlanningInfoForNode()` is needed as it is not initialized.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+    for node in nodes:
+        node.mpi_node = MemoryPlanningInfoForNode()
+        node.mpi_node.pred_buffers = []
+        for dep_name in {dep.name for dep in node.unmet_dependencies}:
```
I used `sorted()` here. To add a bit more context. I used set comprehension here because I noticed that sometimes the different `dep`s could have the same time, I guess due to aliasing. (adding a comment here as well)

> I need to add a linter 

haha.

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+        node.mpi_node = MemoryPlanningInfoForNode()
+        node.mpi_node.pred_buffers = []
+        for dep_name in {dep.name for dep in node.unmet_dependencies}:
+            sched_buf = name_to_buf.get(dep_name)
```
Very good question. Looked around in `scheduler.py` and I think this [line](https://github.com/pytorch/pytorch/blob/b7a5c7d331b2de2470eca4ce45105369320d24ba/torch/_inductor/scheduler.py#L2243) suggests that it should always be in the dictionary, removing the check.

eellison replied:
Sounds good - would be great to remove where not needed

xuanzhang816 commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+            dep.name for dep in node.read_writes.reads - node.unmet_dependencies
+        }:
+            input_buf = name_to_input_buf.get(dep_name)
+            if input_buf:
```
Some of the dependencies are inputs to the program (i.e., buffers with name such as `primals_1` or those aliasing inputs) and thus are not freeable input buffers. Those won't be `name_to_input_buf`, let me also add a comment here.

eellison replied:
add comment inline ?

xuanzhang816 commented:
> @xuanzhang816 in case you're not aware, we also have a dashboard that you can use to get numbers. `bunny pt2dash` . To submit a run of your own go to: https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml . In some cases you'll want to rebase to a commit that has been recently benchmarked.
> 
> There's more details on https://pytorch.org/docs/stable/torch.compiler_performance_dashboard.html. local works as well but compile time on dashboard can be more stable and requires a bit less work locally.

@eellison To do this, I will need to setup ghstack, right?

xuanzhang816 commented:
> @eellison I am addressing the comments locally one by one and resolving them here to keep track of which one I have done. I will create a new PR after I go through everything.

I submitted the restructured code here -- https://github.com/pytorch/pytorch/pull/137205

eellison commented on torch/_inductor/memory.py:
```diff
@@ -0,0 +1,555 @@
+        selected_node = min(
+            nodes_to_schedule,
+            key=lambda node: (
+                max(live_memory + node.mpi_node.size, max_memory),
+                node.mpi_node.size - node.mpi_node.memory_to_free,
+                node.mpi_node.index,
```
Surprising to me that the greedy algorithm is not to order by net change in memory.. This will choose the largest node that is below max memory.. then net difference.. would assume it would just order by net difference.

xuanzhang816 replied:
> I guess the reason this isn't as bad is because it's O(n) * O(max_concurrently_schedulable_nodes).

yes!!

> If `live_memory + node.mpi_node.size  < max_memory`, does it actually matter what we schedule?

In this case, whichever you schedule would not affect the peak memory immediately at the current step, but can have an effect on later steps. 

> I'm not sure we need the max here: `max(live_memory + node.mpi_node.size, max_memory)`.

The algorithm (given in the paper) selects the next node to schedule based on two numbers for every node:

(1) memory after output allocation, call it `mem1`, which is `live_memory + node.mpi_node.size`; 
(2) memory after the node is executed and its last used buffers are freed, call it `mem2`. 

Among all nodes **whose `mem1` is below `max_memory`**, it selects the node with the lowest `mem2`; and if such nodes do not exist, it selects the node with lowest `mem1`. 

The first element in the tuple -- `max(live_memory + node.mpi_node.size, max_memory)` -- is evaluated to be `max_memory` for all nodes **whose `mem1` is below `max_memory`**. For tiebreaking, we then use the second element in the tuple, which is effectively `mem2`: 

`mem2` is `live_memory + node.mpi_node.size - node.mpi_node.memory_to_free`. Since `live_memory` is the same for all nodes, we simply use `node.mpi_node.size - node.mpi_node.memory_to_free` for comparison.

Hope this makes sense.
